src/codegen/translate-expression.lisp

```diff
@@ -145,27 +145,46 @@ Returns a `node'.")
 
              (from-int-method (util:find-symbol "FROMINT" classes-package))
 
+             (val (tc:node-integer-literal-value expr))
+
              (ty (tc:qualified-ty-type qual-ty)))
 
         ;; Avoid casting INTEGER to INTEGER with FROMINT at runtime
-        (if (and (tc:tycon-p ty) (eq 'coalton:integer (tc:tycon-name ty)))
-            (make-node-literal
-             :type tc:*integer-type*
-             :value (tc:node-integer-literal-value expr))
-            (make-node-application
-             :type (tc:qualified-ty-type qual-ty)
-             :rator (make-node-variable
-                     :type (tc:make-function-type*
-                            (list
-                             (pred-type num-pred env)
-                             tc:*integer-type*)
-                            (tc:qualified-ty-type qual-ty))
-                     :value from-int-method)
-             :rands (list
-                     (resolve-dict num-pred ctx env)
-                     (make-node-literal
-                      :type tc:*integer-type*
-                      :value (tc:node-integer-literal-value expr))))))))
+        (cond ((and (tc:tycon-p ty) (eq 'coalton:integer (tc:tycon-name ty)))
```

Comment:
put the otherwise case in an flet called `make-full-call` and do something like

```
(if (not (tyconp ty))
    (make-full-call)
    (case (tycon-name ty)
      (coalton:integer
       (make-node-literal ...))
      ...
      (otherwise
       (make-full-call))))
```